design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,429 @@
+* Kyverno - 12
+* Istio - 14
+* Argo Workflows - 8
+
+Installing too many CRDs also affects the performance of Kubernetes clients like
+kubectl, Helm, and ArgoCD. These clients are often built under the assumption
+that there won’t be more than 50-100 CRDs installed on a cluster.
+
+You can read more about the issues Kubernetes faces when too many CRDs are
+present at:
+
+* https://github.com/crossplane/crossplane/blob/v1.11.2/design/one-pager-crd-scaling.md
+* https://blog.upbound.io/scaling-kubernetes-to-thousands-of-crds/
+
+Apart from performance issues, some folks have expressed concerns about the
+security and general “bloat” of having many unused CRDs installed. I believe the
+security concerns to be an educational issue - I haven't yet seen a convincing
+argument for removing CRDs as opposed to using RBAC.
+
+Crossplane maintainers have invested in improving the performance of the
+Kubernetes control plane and clients in the face of many CRDs. These
+improvements have been insufficient to alleviate community pain. The assumption
+that there won’t be “too many” CRDs is pervasive in the Kubernetes codebase.
+There’s no one easy fix, and the conservative speed at which cloud providers and
+their customers pick up new Kubernetes releases means it can take years for
+fixes to be deployed.
+
+Allowing folks to lower the ratio of installed-to-used CRDs has been proposed
+for some time. The Crossplane maintainers are now aligned that it’s necessary.
+We have explored the following options:
+
+* Make it possible to filter unnecessary CRDs by configuring a provider.
+* Break the biggest providers up into smaller, service-scoped providers.
+* Load CRDs into the API server lazily, on-demand.
+
+## Proposal
+
+I propose that provider maintainers break the largest providers into smaller,
+service-scoped providers. Each service-scoped provider would correspond to what
+is currently an API group within a provider. For example,
+`upbound/provider-aws-rds` would extend Crossplane with support for the 22
+managed resources (and thus 22 CRDs) in the `rds.aws.upbound.io` API group.
+
+### Overview
+
+I proposed we only break up the largest providers. For example:
+
+* `upbound/provider-aws` - Becomes ~150 smaller providers.
+* `upbound/provider-azure` - Becomes ~100 smaller providers.
+* `upbound/provider-gcp` - Becomes ~75 smaller providers.
+* `crossplane-contrib/provider-tencentcloud` - Becomes ~50 smaller providers.
+
+It's important to note that we don't expect anyone to be installing and
+operating all of these providers. Per [Community Impact](#community-impact), we
+expect folks to install ~10 providers on average.
+
+These providers would:
+
+* Be built from “monorepos” - the same repositories they are today, like
+  upbound/provider-aws.
+* Share a `ProviderConfig` by taking a dependency on a (e.g.)
+  `provider-aws-config` package.
+* Continue to use contemporary “compiled-in” cross-resource-references, short to
+  medium term.
```

Comment:
Kind of.

The thinking is that we'd build them all together from the same monorepo, and release them all at once with the same version. So `provider-aws-eks:v0.40.0` and `provider-aws-ec2:v0.40.0` would be released at the same time.

In practice you could almost certainly run `provider-aws-eks:v0.42.0` with `provider-ec2:v0.40.0` without any issue. This will only become more true as we get rid of compiled-in cross-resource-references (see the Cross Resource References heading for more).

That said, running your set of providers at the same version would be the most conservative approach. We probably wouldn't integration test all combinations of provider versions for example - we'd just run our integration tests (that often test across service boundaries) with a set of providers at the same version.